### DISCUSSION TITLE
Harding time grasping the Next.js 13 App Router 'use client' directive

### DISCUSSION THREAD

**Discussion by hbatalhaStch** [vercel/next.js#50864] [Help] [answered]
### Summary

Today I upgraded to nextjs 13 and began getting familiar with it and getting to know its features.
The doc says that whenever I use Use State and Lifecycle Effects I need to use the `'use client'` directive which if I understand it correctly it signals react that the component should be rendered in the browser (client side).

In my current setup using `pages` folder I use `useState` and `useEffect` in practically every file, which means that by moving to the app folder I have to use the `use client` directive making all the pages in my application CSR, correct?

How should I solve this issue? Am I missing something?

**Comment by icyJoseph** [comment, marked as answer]
> it signals react that the component should be rendered in the browser (client side).

That's already not correct. YouTubers and code influencers have spread that concept. It is actually, nuanced. 

- https://nextjs.org/docs/getting-started/react-essentials#client-components

```
In Next.js, they are pre-rendered on the server and hydrated on the client. 
You can think of Client Components as how components in the Pages Router have always worked.
```

>  use client directive making all the pages in my application CSR, correct?

And that's the assumption I've seen many fall into. For example, trying to access window, document, DOM APIs, etc, which inevitable fail during SSR.

There's also this bit in the React Essentials part:

```
Components in the Client Component module graph are primarily rendered on the client, 
but with Next.js, they can also be pre-rendered on the server and hydrated on the client.
```

So this tells me that React Server Components spec, allows what you describe, or rather is not specific about it, so Next.js makes a choice to render Client Components as part of any SSR it does.

**Reply by hbatalhaStch** [reply]
Thanks for clarification, really helpful.
So what happens if I just use `'use client`' in `app/layout.tsx`?

**Reply by icyJoseph** [reply]
Then the entire React tree, or application React tree, will be hydrated client side, that is to say, we need JavaScript that outputs React elements that tells React how the entire page should look like.

> Everything is still, one way or another, part of the SSR process, unless hidden behind interactivity triggers (useEffect, user actions, etc)

That is to say that hydration will touch every single component. So the JS payload will be larger than if you had isolated client components to the leaves. 

Say you have a page with a bunch of buttons (the same button).

If the entire layout is a client component (with `use client`), then there'll be JavaScript that tells React how to generate the layout, page and button React Elements.

When the HTML for the page arrives to the browser, fully SSR'd, JS is downloaded, we'll download JS that outputs the React elements for the layout, page and button. The page contents are visible, but the page is not interactive yet, React starts hydration, where it re-renders the entire tree client side, using the JS it downloaded.

If you had isolated the client components to the button, then only JavaScript to collect the React Elements of those buttons is needed.

After the HTML arrives to the browser, fully SSR'd, JS is downloaded, we'll only download button JS. The page contents are visible, but the page is not interactive yet. React starts hydration, but this time it knows to hydrate only the buttons, and its done.

We get, less download time, less parsing time, and less execution time.

Extreme oversimplification, but this is part of a bigger whole, including streaming and selective hydration, and how React knows where there's slots for server components and client components, depending on where it is rendering.

**Reply by icyJoseph** [reply]
It is worth pointing out that, entire React tree hydration has been the norm. That is what pages dir does. If you ever made a static HTML page with pages dir, and made changes to the HTML, you'd see that they "disappear", but it is just the hydration step, taking over. This is considered though, mismatched HTML.

**Reply by hbatalhaStch** [reply]
Thank you for the great explanation.
